lua/luasnip/extras/snippet_list.lua

```diff
@@ -0,0 +1,75 @@
+local function get_context(snippet)
+	return {
+		name = snippet.name,
+		trigger = snippet.trigger,
+		description = snippet.dscr,
+		wordTrig = snippet.wordTrig and true or false,
+		regTrig = snippet.regTrig and true or false,
+		docstring = snippet.docstring,
+	}
+end
+
+local function available()
+	local luasnip = require("luasnip")
+	local utils = require("luasnip.util.util")
+	local filetypes = utils.get_snippet_filetypes()
+
+	local res = {}
+
+	for _, ft in ipairs(filetypes) do
+		res[ft] = {}
+
+		local snippets = luasnip.get_snippets(ft)
+		for _, snippet in ipairs(snippets) do
+			if not snippet.invalidated then
+				table.insert(res[ft], get_context(snippet))
+			end
+		end
+
+		local autosnippets = luasnip.get_snippets(ft, { type = "autosnippets" })
+		for _, snippet in ipairs(autosnippets) do
+			if not snippet.invalidated then
+				table.insert(res[ft], get_context(snippet))
+			end
+		end
+	end
+
+	return res
+end
+
+local function open_snippet_list()
+	-- load snippets before changing windows/buffers
+	local snippets = available()
+
+	-- create and open buffer on right vertical split
+	vim.cmd("botright vnew")
+
+	local buf = vim.api.nvim_get_current_buf()
+	local win = vim.api.nvim_get_current_win()
+
+	vim.api.nvim_buf_set_name(buf, "Snippets")
+
+	-- make scratch buffer
+	vim.api.nvim_buf_set_option(buf, "filetype", "lua")
+	vim.api.nvim_buf_set_option(buf, "buftype", "nofile")
+	vim.api.nvim_buf_set_option(buf, "bufhidden", "wipe")
+	vim.api.nvim_buf_set_option(buf, "buflisted", false)
+	vim.api.nvim_buf_set_option(buf, "swapfile", false)
+	vim.api.nvim_buf_set_option(buf, "modified", false)
+	vim.api.nvim_buf_set_option(buf, "modeline", false)
+
+	-- disable diagnostics
+	vim.diagnostic.disable(buf)
+
+	-- dump snippets
+	local replacements = vim.split(vim.inspect(snippets), "\n")
+	vim.api.nvim_buf_set_lines(buf, 0, 0, false, replacements)
+
+	-- enable local folding
+	vim.wo[win].foldmethod = "expr"
+	vim.wo[win].foldexpr = [[getline(v:lnum) =~ '= { {' ? 0 : 1]]
```

Comment:
As I said in my answer, maybe manual folding will work better?
Additionally, some other/configurable way to pretty-print the snippets would be a cherry on top.

Reply:
So it turns out that setting the fold method to indent seems to work beautifully. Changes are in https://github.com/L3MON4D3/LuaSnip/pull/652/commits/fe3d9e29cce24af43a5aa0d3bc521c17ebf7b54f.

> Additionally, some other/configurable way to pretty-print the snippets would be a cherry on top.
I'm not entirely sure what you mean by this, would you mind expanding on that? You mean like a modifiable get_context function? Instead of returning a plain object we would return something else depending on parameters passed?

What is pretty-print to you?

Reply:
> I'm not entirely sure what you mean by this, would you mind expanding on that?
> What is pretty-print to you?

Sure! So we have this big list of snippet-types -> filetypes -> these small context-tables, and currently use `vim.inspect` to turn that big table into a readable string, which is then put into the buffer.
`vim.inspect` includes `{`, `}`, `,`, <other characters> to turn its output into almost-valid lua, which we don't really need in our representation, we just want these context-tables grouped by snippet-type and filetype.
This could be achieved by writing a different function, which walks through the big table, and "pretty-prints" each table (so do something slightly different for each level)

For example, the docstrings don't really need quotes, as included by `vim.inspect`, and multiline-strings are tables, and thus look like `{"line1", "line2"}`, which could be improved.

I think it would be best to turn this into a configurable option, with `vim.inspect` as the default.
(we can include other pretty-printers in this module later on)

Reply:
> we just want these context-tables grouped by snippet-type and filetype

So ok let me make sure I'm understanding this correctly.
At the moment we have something like this:

```lua
{
    filetype = {
        snippet_context_table,
        snippet_context_table,
        etc...
    },
    filetype = {
        snippet_context_table,
        snippet_context_table,
        etc...
    },
}
```
and you're visualizing something more like this?

```lua
{
    filetype = {
        snippet_type = {
            snippet_context_table,
            etc...
        },
        snippet_type = {
            snippet_context_table,
            etc...
        },
    },
}
```

>  to turn its output into almost-valid lua, which we don't really need in our representation
> the docstrings don't really need quotes

So you're not visualizing this as a 'lua' type buffer? Just a text file?

> This could be achieved by writing a different function, which walks through the big table, and "pretty-prints" each table (so do something slightly different for each level)

So this function would be working on the context tables?

> multiline-strings are tables, and thus look like {"line1", "line2"}, which could be improved

Can you show me an example of where this happens?

>  this into a configurable option, with vim.inspect as the default.
(we can include other pretty-printers in this module later on)

So similar to the idea with configuring the buffer name, we could include a ```printer``` value in an opts table which is a function that grabs the big table that we get back from the ```available()``` function and turns it into the string that we want to output if I'm understanding you correctly.

I'm trying my best to understand what you're visualizing but I think I'm still confused I'm so sorry haha
Would it be possible for you to draw/sketch/pseucode what the output is supposed to look like in your eyes, I think that would help me make things clearer in my mind.

Sorry about that.

Reply:
> > we just want these context-tables grouped by snippet-type and filetype
> 
> So ok let me make sure I'm understanding this correctly. At the moment we have something like this:
> 
> ```lua
> {
>     filetype = {
>         snippet_context_table,
>         snippet_context_table,
>         etc...
>     },
>     filetype = {
>         snippet_context_table,
>         snippet_context_table,
>         etc...
>     },
> }
> ```
> 
> and you're visualizing something more like this?
> 
> ```lua
> {
>     filetype = {
>         snippet_type = {
>             snippet_context_table,
>             etc...
>         },
>         snippet_type = {
>             snippet_context_table,
>             etc...
>         },
>     },
> }
> ```

Ah, disregard that, I thought the table was already grouped by snippet-type (regular vs autosnippet). This distinction might be nice to have, but it's not vital right now.

> > to turn its output into almost-valid lua, which we don't really need in our representation
> > the docstrings don't really need quotes
> 
> So you're not visualizing this as a 'lua' type buffer? Just a text file?

Yes, exactly. Similar to nvim-tree, but not of files and directories, but of snippets and file/snippet-types.

> > This could be achieved by writing a different function, which walks through the big table, and "pretty-prints" each table (so do something slightly different for each level)
> 
> So this function would be working on the context tables?

:+1:

> > multiline-strings are tables, and thus look like {"line1", "line2"}, which could be improved
> 
> Can you show me an example of where this happens?

Oh, `get_docstring` should return a `string[]`, weird that it doesn't show up that way..

> > this into a configurable option, with vim.inspect as the default.
> > (we can include other pretty-printers in this module later on)
> 
> So similar to the idea with configuring the buffer name, we could include a `printer` value in an opts table which is a function that grabs the big table that we get back from the `available()` function and turns it into the string that we want to output if I'm understanding you correctly.

Yes! default-value can be just `vim.inspect` since it has the correct arguments+output already.

> I'm trying my best to understand what you're visualizing but I think I'm still confused I'm so sorry haha Would it be possible for you to draw/sketch/pseucode what the output is supposed to look like in your eyes, I think that would help me make things clearer in my mind.

For example
```
lua
 ├─trigger: adsf
 │ docstring: adsf
 │ ...
 │ ...
 └─trigger: asdf2
   docstring: adsf2

python
 ├─trigger: adsf
 │ docstring: adsf
 │ ...
 │ ...
 └─trigger: asdf2
   docstring: adsf2
   ...
   ...
```
would look pretty cool

It's not necessary to include some advanced pretty-printer from the get-go, but implementing the configurability now would be great, then it's easy to improve later on.